### PR TITLE
feat: make _canonical_matrix faster

### PR DESCRIPTION
**Pull request by thofma**
Some low-hanging fruit.

Before:
```
julia> @time Oscar.Orderings._canonical_matrix(w);
  1.322101 seconds (21.72 M allocations: 826.664 MiB, 14.43% gc time)
```
After:
```
julia> @time Oscar.Orderings._canonical_matrix(w);
  0.021654 seconds (3.57 k allocations: 2.066 MiB)
```
There is still some room for improvement.

### CODE REVIEW

**Comment by HechtiDerLachs** [comment]
Cool, thanks! Definitely an improvement. But as you say: Still room for further improvement, eventually.

**Review by HechtiDerLachs — DISMISSED** [review]

File: src/Rings/orderings.jl [hunk 1]
```diff
@@ -88,30 +88,35 @@ struct MatrixOrdering <: AbsGenOrdering
...
     if !isone(c)
-      nw = divexact(nw, c)
+      nw = divexact!(nw, nw, c)
     end
-    for j in 1:nrows(ww)
-      h = findfirst(x->ww[j, x] != 0, 1:ncols(w))
-      if !iszero(nw[1, h])
-        nw = abs(ww[j, h])*nw - sign(ww[j, h])*nw[1, h]*ww[j:j, :]
+    for j in 1:k
+      h = findfirst(x-> is_zero_entry(ww, j, x), 1:ncols(w))
```
Comment (lgoettgens) [inline src/Rings/orderings.jl]: findfirst can return `nothing`. Could you add a check for this?

**Comment by thofma** [comment]
I guess I messed something up.

How fast is your sparse version?

**Comment by HechtiDerLachs** [comment]
> How fast is your sparse version?

My sparse version never became functional, unfortunately. I also messed things up and didn't know how to resolve it.

File: src/Rings/orderings.jl (hunk 1 above)
Comment (thofma) [reply src/Rings/orderings.jl]: It should never be `nothing` and would be catched later anyway.

**Comment by joschmitt** [comment]
I pushed a fix for the sparse version. I now get these timings:
```
# This PR:
julia> @btime Oscar.Orderings._canonical_matrix(M);
  9.841 ms (4095 allocations: 2.09 MiB)

# 4114:
julia> @btime Oscar.Orderings._canonical_matrix(M);
  5.179 ms (260187 allocations: 4.11 MiB)
```

EDIT: new version of #4114
```
julia> @btime Oscar.Orderings._canonical_matrix(M);;
  1.505 ms (6669 allocations: 247.57 KiB)
```

File: src/Rings/orderings.jl [hunk 2]
```diff
@@ -116,6 +116,43 @@ function _canonical_matrix(w)
...
+  piv = Int[]
+  # piv[i] stores the column index of the first non-zero entry in row i of ww
+  for i in 1:nrows(w)
+    if is_zero_row(w, i)
+      continue
+    end
+    nw = view(w, i:i, :)
+    c = content(nw)
+    if !isone(c)
+      nw = divexact!(nw, nw, c)
```
Comment (joschmitt) [inline src/Rings/orderings.jl]: `divexact!` returns `nothing`, so this sets `nw` to `nothing` (if it is called).

File: src/Rings/orderings.jl [hunk 3]
```diff
@@ -87,7 +87,7 @@ struct MatrixOrdering <: AbsGenOrdering
   end
 end
 
-function _canonical_matrix(w)
+function __canonical_matrix(w)
```
Comment (lgoettgens) [inline src/Rings/orderings.jl]: this is no longer used and should be removed (before merging this)

**Review by lgoettgens — CHANGES_REQUESTED** [review]

File: src/Rings/orderings.jl (hunk 2 above)
Comment (fingolfin) [reply src/Rings/orderings.jl]: Hopefully will be fixed by https://github.com/Nemocas/Nemo.jl/pull/1858

**Review by fingolfin — APPROVED** [review]

**Review by lgoettgens — APPROVED** [review]

**Pull request merged**